Sources/MongoSwift/MongoClient.swift

```diff
@@ -786,7 +907,9 @@ public class MongoClient {
      * to continue to receive events.
      */
     public func addCommandEventHandler<T: CommandEventHandler>(_ handler: T) {
-        self.commandEventHandlers.append(WeakEventHandler<T>(referencing: handler))
+        self.eventHandlerLock.withLock {
```

Comment:
related to my comment above re the lock (not something we need to fix today, but for the future): with the number of places in the code we now guard some value behind a lock and just have to remember that it should only be accessed via the lock, it would be a nice improvement to introduce some generic wrapper type like `Locked<T>` that stores a value of type `T` plus its corresponding `Lock` and has a method that lets you do something like
```swift
myLockedThing.withLockedValue { value in
    // ... do something with value
}
```
The method would acquire the lock, execute your provided closure with the value it guards, then releases the lock. the locked value could be private to the `Locked` type so it's impossible to access it other than when you hold the lock, which would make it harder for us to forget to acquire the lock. 

I filed SWIFT-1612 about this. it could be a nice little ticket to pick up if you end up having any free time waiting for reviews in the future